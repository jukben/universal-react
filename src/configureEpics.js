import { combineEpics } from 'redux-observable';
import { epics as viewer } from './ducks/viewer';

const epics = [...viewer];

export default (deps = {}, platformEpics = []) => (action$, { getState }) =>
  combineEpics(...epics, ...platformEpics)(action$, { ...deps, getState });
