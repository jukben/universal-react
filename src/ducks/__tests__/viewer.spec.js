import configureMockStore from 'redux-mock-store';
import { createEpicMiddleware } from 'redux-observable';
import viewerReducer, { fetchInitAction, fetchSuccessAction } from '../viewer';
import configureEpics from '../../configureEpics';

import stateFetchedMock from '../../__test__/state.fetched.json';
import httpMockSuccess from '../../api/__tests__/mock.successFetch.json';
import httpMockError from '../../api/__tests__/mock.errorAuth.json';

const wait = (time = 0) => new Promise(res => setTimeout(() => res(), time));

describe('Viewer reducer', () => {
  it('fetch init should set loading to true', () => {
    expect(
      viewerReducer(
        {
          searchLoading: false,
          lastError: null,
        },
        fetchInitAction(),
      ),
    ).toEqual({ lastError: null, searchLoading: true });
  });

  it('fetch success should set loading to false', () => {
    expect(
      viewerReducer(
        {
          searchLoading: false,
          lastError: null,
        },
        fetchSuccessAction(),
      ),
    ).toEqual({ lastError: null, searchLoading: false });
  });
});

const rootEpic = configureEpics();
const epics = createEpicMiddleware(rootEpic);
const mockStore = configureMockStore([epics]);

describe('Viewer epics', () => {
  let store;

  beforeEach(() => {
    store = mockStore(stateFetchedMock);
  });

  afterEach(() => {
    store.clearActions();
    epics.replaceEpic(rootEpic);
    fetch.resetMocks();
  });

  it('should correctly dispatch fetch success', () => {
    store.dispatch(fetchSuccessAction('random joke haha'));

    expect(store.getActions()).toMatchSnapshot();
  });

  it('should correctly fetch random gif', async () => {
    fetch.mockResponse(JSON.stringify(httpMockSuccess), { status: 200 });
    store.dispatch(fetchInitAction());

    await wait();
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should make correct report about error', async () => {
    fetch.mockResponse(JSON.stringify(httpMockError), { status: 400 });
    store.dispatch(fetchInitAction());

    await wait();
    expect(store.getActions()).toMatchSnapshot();
  });
});
