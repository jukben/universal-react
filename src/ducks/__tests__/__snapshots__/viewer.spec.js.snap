// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer epics should correctly dispatch fetch success 1`] = `
Array [
  Object {
    "payload": Object {
      "joke": "random joke haha",
    },
    "type": "universalreact/viewer/FETCH_SUCCESS",
  },
]
`;

exports[`Viewer epics should correctly fetch random gif 1`] = `
Array [
  Object {
    "type": "universalreact/viewer/FETCH_INIT",
  },
  Object {
    "payload": Object {
      "joke": "Chuck Norris has seen it all, and done it all, so many times that he has long ago lost count.",
    },
    "type": "universalreact/viewer/FETCH_SUCCESS",
  },
]
`;

exports[`Viewer epics should make correct report about error 1`] = `
Array [
  Object {
    "type": "universalreact/viewer/FETCH_INIT",
  },
  Object {
    "payload": Object {
      "error": Object {
        "error": true,
        "error_description": "ChuckNorris.io API error",
      },
    },
    "type": "universalreact/viewer/FETCH_ERROR",
  },
]
`;
